src/tracks/SampleTrack.cpp

```diff
@@ -573,57 +574,104 @@ SampleTrack::~SampleTrack()
 
 
 bool SampleTrack::play( const MidiTime & _start, const fpp_t _frames,
-					const f_cnt_t _offset, int _tco_num )
+						const f_cnt_t _offset, int _tco_num )
 {
 	m_audioPort.effects()->startRunning();
-	bool played_a_note = false;	// will be return variable
+
+	//will be return variable
+	bool played_a_note = false;
 
 	tcoVector tcos;
+
 	::BBTrack * bb_track = NULL;
+	MidiTime bbEndPos = 0;
+
+	/********************************************************************
+	* In this whole block we find our BB Track TCO in the Songeditor
+	* and so we can calculate the length of it.
+	* If _tco_num is equal or bigger than 0, we know we have a
+	* Sampletrack within a BB-Container
+	********************************************************************/
+
 	if( _tco_num >= 0 )
 	{
-		if( _start != 0 )
-		{
-			return false;
-		}
-		tcos.push_back( getTCO( _tco_num ) );
-		if (trackContainer() == (TrackContainer*)Engine::getBBTrackContainer())
+		//the BB-Track in Songeditor
+		bb_track = BBTrack::findBBTrack( _tco_num );
+		//Vector of BB Tcos in Songeditor. It will be filled later
+		tcoVector bbtcos;
+		//Do we play in BB-Editor or in Songeditor?
+		if( Engine::getSong()->playMode() != Song::Mode_PlayBB )
 		{
-			bb_track = BBTrack::findBBTrack( _tco_num );
+			MidiTime currentSongPos = Engine::getSong()->getPlayPos( Song::Mode_PlaySong ).getTicks();
+			bb_track->getTCOsInRange( bbtcos, currentSongPos, currentSongPos );
+			if( bbtcos.size() != 0 )
+			{
+				for( int i = 0; i < bbtcos.size(); ++i )
```

Comment:
I think it'd be better to use range-based for loop. Also, I don't think the size check is needed.